### PR TITLE
Youtube: add plugin for viewing videos/channels

### PR DESCRIPTION
**Pull request by mia-pi-git**
This is mostly for the youtube room. However, /requestapproval and /link are meant for all rooms, since they're very commonly used through Kid A and very useful, and can be streamlined a bit here (they're in this plugin so they can make use of the Youtube interface.) 
Specifics are here https://www.smogon.com/forums/threads/youtube-channel-database.3664726/.
This has been approved by both Kie and A Random Duck.
As for the Config setting for the api key, i have a key that can be used, and can share privately for someone who can set that.

### CODE REVIEW

File: server/chat-plugins/youtube.ts [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+import * as https from 'https';
+import {FS} from '../../lib/fs';
+const ROOT = 'https://www.googleapis.com/youtube/v3/';
+const CHANNEL = `${ROOT}channels`;
+const STORAGE_PATH = 'config/chat-plugins/youtube.json';
+
+let channelData: AnyObject;
+
+try {
+	channelData = JSON.parse(FS(STORAGE_PATH).readSync()) || {};
```
Comment (KrisXV) [inline server/chat-plugins/youtube.ts:17]:
```suggestion
	channelData = JSON.parse(FS(STORAGE_PATH).readIfExistsSync() || "{}");
```

just to keep consistent with the other files

File: server/chat-plugins/youtube.ts [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+} catch (e) {
+	channelData = {};
+}
+
+export const YouTube = new class {
+	interval: NodeJS.Timeout | null;
```
Comment (KrisXV) [inline server/chat-plugins/youtube.ts:23]: wouldnt this be better suited as `NodeJS.Timer`?

File: server/chat-plugins/youtube.ts [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+
+let channelData: AnyObject;
+
+try {
+	channelData = JSON.parse(FS(STORAGE_PATH).readSync()) || {};
+} catch (e) {
+	channelData = {};
+}
+
+export const YouTube = new class {
```
Comment (KrisXV) [inline server/chat-plugins/youtube.ts:22]:
```suggestion
export class Youtube {
```

File: server/chat-plugins/youtube.ts [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+		if (username) {
+			buf += `<p style="text-align:left;font-style:italic;color:white;">PS username: ${username}</p></td></tr></table></div>`;
+		} else {
+			buf += '</td></tr></table></div>';
+		}
+		return buf;
+	}
+	randChannel() {
+		const keys = Object.keys(channelData);
+		const id = keys[Math.floor(Math.random() * keys.length)].trim();
```
Comment (KrisXV) [inline server/chat-plugins/youtube.ts:79]:
```suggestion
		const id = Dex.shuffle(keys)[0].trim();
```

(this is optional, i just think that if we have a function, why not use it. though then it brings the question of why `sample` is on the `PRNG` class and not on `Dex` as well)

File: server/chat-plugins/youtube.ts (hunk 2 above)
Comment (mia-pi-git) [reply server/chat-plugins/youtube.ts:23]: I don't think so?
VSCode says setInterval is `NodeJS.Timeout`.

**Comment by mia-pi-git** [comment]
> I personally think the requestapproval/link function that is ported from Kid A should be in a separate commit.

I felt it fit here, since it uses `YoutubeInterface` functions, but if it's desired, I can move it out of this PR.

File: server/chat-plugins/youtube.ts (hunk 2 above)
Comment (xfix) [reply server/chat-plugins/youtube.ts:23]: It doesn't matter. Checking `@types/node`'s source code, `Timeout` is a concrete class that implements `Timer` interface.

Comment (mia-pi-git) [reply server/chat-plugins/youtube.ts]: Alright.

**Comment by mia-pi-git** [comment]
(requesting since they ran the plugin on the bot, so i thought input from them might be helpful, lmk if I shouldn't have done that)

**Comment by mia-pi-git** [comment]
Removed approvals/link for now.

**Comment by Zarel** [comment]
Looks good to me!

**Pull request merged**